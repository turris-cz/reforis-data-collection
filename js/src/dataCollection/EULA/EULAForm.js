/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { RadioSet } from "foris";
import React from "react";

export default function EULAForm({
    formData, setFormValue, formErrors, onModalToggle, ...props
}) {
    return (
        <>
            <RadioSet
                choices={getEULAChoices(onModalToggle)}
                name={_("Agreement")}
                value={formData.eula.toString()}
                onChange={
                    setFormValue((value) => ({ eula: { $set: value } }))
                }
                {...props}
            />
        </>
    );
}

function getEULAChoices(onModalToggle) {
    function onClickHandler(e) {
        // `target` is the element the click was on (the div we hooked or an element
        // with in it), `currentTarget` is the div we hooked the event on
        const el = e.target.closest("a");
        if (el && e.currentTarget.contains(el)) {
            e.preventDefault();
            onModalToggle();
        }
    }

    return [
        {
            value: "0",
            // eslint-disable-next-line max-len
            // eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions
            label: <div
                onClick={onClickHandler}
                dangerouslySetInnerHTML={{
                    __html:
                    _("I do not accept the <a href=\"#\">Terms of Participation in Turris Project (Data Collection)</a>."),
                }}
            />,
        },
        {
            value: "1",
            // eslint-disable-next-line max-len
            // eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions
            label: <div
                onClick={onClickHandler}
                dangerouslySetInnerHTML={{
                    __html:
                    _("I accept the <a href=\"#\">Terms of Participation in Turris Project (Data Collection)</a>."),
                }}
            />,
        },
    ];
}
