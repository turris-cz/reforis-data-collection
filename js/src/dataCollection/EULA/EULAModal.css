@media (min-width: 576px) {
    .modal-dialog {
        max-width: 60rem;
    }
}
