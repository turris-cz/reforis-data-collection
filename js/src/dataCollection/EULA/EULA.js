/*
 * Copyright (C) 2020-2021 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useState } from "react";
import PropTypes from "prop-types";

import EULAModal from "./EULAModal";
import EULAForm from "./EULAForm";

import "./EULA.css";

EULA.propTypes = {
    formData: PropTypes.object,
    setFormValue: PropTypes.func,
    disabled: PropTypes.bool,
};

export default function EULA({ formData, setFormValue, disabled }) {
    const [shown, setShown] = useState(false);

    function onModalToggle() {
        setShown((currentShown) => !currentShown);
    }

    return (
        <>
            <EULAForm
                formData={formData}
                setFormValue={setFormValue}
                disabled={disabled}
                onModalToggle={onModalToggle}
            />
            <EULAModal formData={formData} shown={shown} setShown={setShown} />
        </>
    );
}
