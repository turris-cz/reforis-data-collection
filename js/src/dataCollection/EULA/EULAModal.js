/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";
import {
    API_STATE, ErrorMessage, Modal, ModalBody, ModalHeader, Spinner, useAPIGet,
} from "foris";

import API_URLs from "API";
import "./EULAModal.css";

EULAModal.propTypes = {
    shown: PropTypes.bool.isRequired,
    setShown: PropTypes.func.isRequired,
    formData: PropTypes.shape({
        eula: PropTypes.number.isRequired,
        token: PropTypes.string,
    }),
};

export default function EULAModal({ shown, setShown, formData }) {
    const { eula } = formData;
    const [getEulaState, getEula] = useAPIGet(API_URLs.eula);

    useEffect(() => {
        getEula({ data: { eula } });
    }, [eula, getEula]);

    let content;
    if (getEulaState.state === API_STATE.INIT || getEulaState.state === API_STATE.SENDING) {
        content = <Spinner />;
    } else if (getEulaState.state === API_STATE.ERROR) {
        content = <ErrorMessage />;
    } else {
        content = (
            <div className="display-linebreak text-monospace">
                {getEulaState.data.text}
            </div>
        );
    }

    return (
        <>
            <Modal
                scrollable
                setShown={setShown}
                shown={shown}
            >
                <ModalHeader
                    title={_("Terms of Participation in Turris Project (Data Collection)")}
                    setShown={setShown}
                />
                <ModalBody>
                    {content}
                </ModalBody>
            </Modal>
        </>
    );
}
