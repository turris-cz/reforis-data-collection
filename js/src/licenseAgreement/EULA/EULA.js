/*
 * Copyright (C) 2020-2025 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useState } from "react";

import PropTypes from "prop-types";

import EULAForm from "./EULAForm";
import EULAModal from "./EULAModal";

import "./EULA.css";

EULA.propTypes = {
    formData: PropTypes.object,
    setFormValue: PropTypes.func,
    disabled: PropTypes.bool,
};

export default function EULA({ formData, setFormValue, disabled }) {
    const [shown, setShown] = useState(false);

    return (
        <>
            <EULAForm
                formData={formData}
                setFormValue={setFormValue}
                disabled={disabled}
                onModalToggle={() => setShown((currentShown) => !currentShown)}
            />
            <EULAModal formData={formData} shown={shown} setShown={setShown} />
        </>
    );
}
