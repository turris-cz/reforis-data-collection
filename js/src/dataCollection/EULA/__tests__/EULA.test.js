/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import mockAxios from "jest-mock-axios";

import diffSnapshot from "snapshot-diff";
import {render, fireEvent, wait, waitForElement, queryAllByText} from "foris/testUtils/customTestRender";

import EULA from '../EULA';


describe("<DataCollection />", () => {
    let getByText;
    let getByLabelText;
    let container;
    let asFragment;

    beforeEach(async () => {
        ({getByText, getByLabelText, container, asFragment} = render(<>
            <EULA/>
            <div id="modal-container"/>
        </>));
        await waitForElement(() => getByText("License Agreement"));
    });

    it("Should request GET setting and GET eula.", async () => {
        expect(mockAxios.get).toBeCalledWith("/reforis/data-collection/api/settings", expect.anything());
        mockAxios.mockResponse({data: {eula: 0, token: "random_token"}});
        await wait(() => {
            expect(mockAxios.get).toBeCalledWith("/reforis/data-collection/api/eula", expect.anything());
        });
    });

    it("Should render form.", async () => {
        mockAxios.mockResponse({data: {eula: 0, token: "random_token"}});
        await wait(() => {
            expect(mockAxios.get).toBeCalledWith("/reforis/data-collection/api/eula", expect.anything());
        });
        mockAxios.mockResponse({data: {version: 1, text: "eula text"}});
        await waitForElement(() => getByLabelText(/I accept/));

        getByLabelText(/I do not accept/);
        getByText("Save");
        expect(container).toMatchSnapshot();
    });

    it("Should toggle modal.", async () => {
        mockAxios.mockResponse({data: {eula: 0, token: "random_token"}});
        await wait(() => {
            expect(mockAxios.get).toBeCalledWith("/reforis/data-collection/api/eula", expect.anything());
        });
        mockAxios.mockResponse({data: {version: 1, text: "eula text"}});
        await waitForElement(() => getByText("eula text"));
        const beforeModalToggle = asFragment();

        fireEvent.click(queryAllByText(container,"Terms of Participation in Turris Project (Data Collection)")[0]);

        expect(diffSnapshot(beforeModalToggle, asFragment())).toMatchSnapshot();
    });
});

