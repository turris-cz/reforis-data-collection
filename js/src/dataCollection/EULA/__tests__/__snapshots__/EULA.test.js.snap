// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataCollection /> Should render form. 1`] = `
<div>
  <h3>
    License Agreement
  </h3>
  <p>
    It's required to confirm the Terms of Participation in Turris Project to participate in data collection.
  </p>
  <div
    class="col-sm-12 offset-lg-1 col-lg-10 p-0 mb-3"
  >
    <form>
      <div
        class="form-group"
      >
        <div
          class="custom-control custom-radio custom-control-inline"
        >
          <input
            checked=""
            class="custom-control-input"
            id="Agreement-0"
            name="Agreement"
            type="radio"
            value="0"
          />
          <label
            class="custom-control-label"
            for="Agreement-0"
          >
            <div>
              I do not accept the 
              <a
                href="#"
              >
                Terms of Participation in Turris Project (Data Collection)
              </a>
              .
            </div>
          </label>
        </div>
        <div
          class="custom-control custom-radio custom-control-inline"
        >
          <input
            class="custom-control-input"
            id="Agreement-1"
            name="Agreement"
            type="radio"
            value="1"
          />
          <label
            class="custom-control-label"
            for="Agreement-1"
          >
            <div>
              I accept the 
              <a
                href="#"
              >
                Terms of Participation in Turris Project (Data Collection)
              </a>
              .
            </div>
          </label>
        </div>
      </div>
      <div
        class="text-right"
      >
        <button
          class="btn btn-primary  col-sm-12 col-lg-3"
          type="submit"
        >
           
           
          Save
        </button>
      </div>
    </form>
  </div>
  <div
    id="modal-container"
  >
    <div
      class="modal fade "
      role="dialog"
    >
      <div
        class="modal-dialog modal-dialog-centered modal-dialog-scrollable"
        role="document"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <h5
              class="modal-title"
            >
              Terms of Participation in Turris Project (Data Collection)
            </h5>
            <button
              class="close"
              type="button"
            >
              <span
                aria-hidden="true"
              >
                Ã—
              </span>
            </button>
          </div>
          <div
            class="modal-body"
          >
            <div
              class="display-linebreak text-monospace"
            >
              eula text
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<DataCollection /> Should toggle modal. 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -78,11 +78,11 @@
    </div>
    <div
      id=\\"modal-container\\"
    >
      <div
-       class=\\"modal fade \\"
+       class=\\"modal fade show\\"
        role=\\"dialog\\"
      >
        <div
          class=\\"modal-dialog modal-dialog-centered modal-dialog-scrollable\\"
          role=\\"document\\""
`;
